libyul/backends/evm/Dominator.h

```diff
@@ -0,0 +1,275 @@
+
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Dominator analysis of a control flow graph.
+ * The implementation is based on the following paper:
+ * https://www.cs.princeton.edu/courses/archive/spr03/cs423/download/dominators.pdf
+ * See appendix B pg. 139.
+ */
+#pragma once
+
+#include <libyul/backends/evm/ControlFlowGraph.h>
+#include <libsolutil/Visitor.h>
+
+#include <range/v3/algorithm.hpp>
+#include <range/v3/view/drop.hpp>
+#include <range/v3/view/reverse.hpp>
+#include <range/v3/view/transform.hpp>
+
+#include <deque>
+#include <map>
+#include <vector>
+#include <set>
+
+namespace solidity::yul
+{
+
+template<typename Vertex, typename ForEachSuccessor>
+class Dominator
+{
+public:
+
+	Dominator(Vertex const& _entry, size_t _numVertices):
+		m_vertex(_numVertices),
+		m_immediateDominator(lengauerTarjanDominator(_entry, _numVertices))
+	{
+		buildDominatorTree();
+	}
+
+	std::vector<Vertex> const& vertices() const
+	{
+		return m_vertex;
+	}
+
+	std::map<Vertex, size_t> const& vertexIndices() const
+	{
+		return m_vertexIndex;
+	}
+
+	std::vector<size_t> const& immediateDominators() const
+	{
+		return m_immediateDominator;
+	}
+
+	std::map<size_t, std::vector<size_t>> const& dominatorTree() const
+	{
+		return m_dominatorTree;
+	}
+
+	// Checks whether ``_a`` dominates ``_b`` by going
+	// through the path from ``_b`` to the entry node.
+	// If ``_a`` is found, then it dominates ``_b``
+	// otherwise it doesn't.
+	bool dominates(Vertex const& _a, Vertex const& _b)
+	{
+		size_t aIdx = m_vertexIndex[_a];
+		size_t bIdx = m_vertexIndex[_b];
+
+		if (aIdx == bIdx)
+			return true;
+
+		size_t idomIdx = m_immediateDominator[bIdx];
+		while (idomIdx != 0)
+		{
+			if (idomIdx == aIdx)
+				return true;
+			idomIdx = m_immediateDominator[idomIdx];
+		}
```

Comment:
> You have very few sanity checks. You're barely using asserts - basically only to check arguments. There are always some invariants or assumptions in such algorithms. It's good to have checks against them.
>
> For example if I'm not mistaken, the immediate dominator will always have a lower index than the vertex it dominates, right? Such a property guarantees that we'll not loop forever here so it would be a good idea to assert that it's true.

Yes. I added some more invariants and assumptions.